aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -126,4 +126,74 @@ Tensor flip_cuda(const Tensor& self, IntList dims) {
   return out_tensor;
 }
 
+template <typename scalar_t>
+__global__
+void roll_cuda_kernel(scalar_t* in_tensor, scalar_t* out_tensor, int64_t N,
+                      int64_t roll_dim, int64_t shift, int64_t start,
+                      int64_t* shape, int64_t total_dims) {
+  int64_t linear_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (linear_index >= N) {
+    return;
+  }
+
+  // amount by which to increment or decrement linear_index to shift that
+  // dimensions by 1.
+  int64_t roll_dim_multiplier = 1;
+  for (int64_t i = total_dims - 1; i  > roll_dim; i--) {
+    roll_dim_multiplier *= shape[i];
+  }
+  // roll dim idx is the index of linear_index along the rolling dimension.
+  int64_t roll_dim_idx = linear_index % (roll_dim_multiplier * shape[roll_dim]) / roll_dim_multiplier;
+  // index into the source data to find appropriate value.
+  int64_t source_idx = 0;
+  if( roll_dim_idx >= (shape[roll_dim] - start) ) {
+    source_idx = linear_index - (shift * roll_dim_multiplier);
+  } else {
+    source_idx = linear_index + (start * roll_dim_multiplier);
+  }
+  out_tensor[linear_index] = in_tensor[source_idx];
+}
+
+// Roll a tensor along a dimension
+Tensor roll_cuda(const Tensor& self, IntList shifts, IntList dims) {
+  // todo: support rolling along no or multiple dimensions as in numpy.roll.
+  AT_CHECK(dims.size() == 1, "only single dimension roll currently supported");
+  AT_CHECK(shifts.size() == dims.size(), "shifts and dimensions must align");
+  // If the first dimension is zero, this is an empty tensor and rolls do nothing.
+  // Return a clone so the caller can safely modify result, and avoid a div by
+  // zero error below.
+  if( self.size(0) == 0 ) {
+    return self.clone();
+  }
+  const int64_t N = self.numel();
+  const int64_t dim = dims[0];
+  const int64_t size = self.size(dim);
+  int64_t start = (size - shifts[0]) % size;
+  // Behavior of % is different in C++ vs Python for negative numbers. This
+  // corrects the difference.
+  if( start < 0 ) start = start + size;
+
+  const int64_t block_size = 512;
+  dim3 dim_block(block_size);
+  dim3 dim_grid((N + block_size - 1) / block_size);
+
+  auto total_dims = self.dim();
+  auto shape = self.sizes().vec();
+  auto shape_t = at::CPU(kLong).tensorFromBlob(shape.data(), {static_cast<int64_t>(shape.size())});
+
+  auto out_tensor = at::empty_like(self);
+  if (out_tensor.numel() == 0) {
+    return out_tensor;
+  }
+
+  AT_DISPATCH_ALL_TYPES_AND_HALF(self.type(), "roll_cuda", [&] {
```

Comment:
I don't see how this can work without passing in the strides as well.

Reply:
Can you add a test case with a non-contiguous tensor to test this?

Reply:
updated the code to convert non-contiguous to contiguous and added a test with a non-contiguous tensor.